src/mblm/train/core/trainer.py

```diff
@@ -833,8 +833,9 @@ def before_new_epoch(epoch: int) -> None:
                 fw_exec_time = time() - start_fw_measure
                 train_loss_as_flt = float(train_loss.item())
                 if math.isnan(train_loss_as_flt):
+                    shape = batch.shape if isinstance(batch, torch.Tensor) else batch[0].shape
```

Comment:
Since there is no constraint on `batch` of type `TBatch` (i.e., `TBatch = TypeVar("TBatch")`), I would rather not assume that it is a list or tuple of tensors. I suggest **either**:
- Expanding this check to make sure that batch is iterable and its first element is a `Tensor`
- Leaving this check and constraining `TBatch` to be `TBatch = TypeVar("TBatch", bound=torch.Tensor | Iterable[torch.Tensor])` and also type `batch` as `batch: TBatch =  iteration.item` in the beginning of each iteration. This is more elegant and compatible with the types of batches that we have used so far (e.g., single Tensor or tuple/list of Tensors) , but maybe there are use cases I have not considered

Reply:
Now I was fully explicit on all the types returned by `epoch_cycler`, please have a look again!